azure-arc-for-servers/grant-permission.ps1

```diff
@@ -0,0 +1,43 @@
+# you can run anywhere
+
+function Get-AzAccountAccessToken()
+{
+    if(-not (Get-Module Az.Accounts)) {
+        Import-Module Az
+    }
+
+    $audience ="https://management.core.windows.net/"       
+    $context = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile.DefaultContext
+    $token = [Microsoft.Azure.Commands.Common.Authentication.AzureSession]::Instance.AuthenticationFactory.Authenticate($context.Account, $context.Environment, $context.Tenant.Id.ToString(), $null, [Microsoft.Azure.Commands.Common.Authentication.ShowDialog]::Never, $null, $audience).AccessToken
+    return $token
+}
+
+function Set-KeyVaultAccessPolicy {
+    [CmdletBinding()]
+    param (
+       [Parameter(Mandatory=$true)]
+       [string] $machineName,
+       [Parameter(Mandatory=$true)]
+       [string] $subscriptionId,
+       [Parameter(Mandatory=$true)]
+       [string] $resourceGroup,
+       [Parameter(Mandatory=$true)]
+       [string] $keyVaultName,
+       [string] $apiVersion ="2019-08-15"
+    )
+    
+    $machineUri= "https://management.azure.com/subscriptions/${subscriptionId}/resourcegroups/${resourceGroup}/providers/Microsoft.HybridCompute/machines/${machineName}?api-version=${apiVersion}"
+
+    # get a token for the target audience. It's the ARM management
+    $yourtoken= Get-AzAccountAccessToken
+  
+    # get the principal id of your connected machine
+    $response1= Invoke-WebRequest -Method GET -Uri $machineUri -UseBasicParsing -Headers @{Metadata="True"; Authorization="bearer $yourtoken"}
```

Comment:
Why are you not using `Invoke-RestMethod`? You get JSON parsing for free. No need to convert the results.

Also, `-UseBasicParsing` should be removed.

Reply:
Since I have to use Invoke-WebRequest in the call, for the consistency, I use invoke-webrequest all in the script.

Reply:
Can you use `Invoke-RestMethod` throughout?